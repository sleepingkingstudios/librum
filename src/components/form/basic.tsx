import * as React from 'react';
import {
  Form as FormWrapper,
  Formik,
  FormikValues,
} from 'formik';
import type { OnSubmit } from './types';

interface IBasicFormProps {
  children: React.ReactNode;
  initialValues?: FormikValues;
  onSubmit: OnSubmit;
}

export const BasicForm = (
  { children, initialValues, onSubmit }: IBasicFormProps
): JSX.Element => {
  return (
    <Formik initialValues={initialValues} onSubmit={onSubmit}>
      <FormWrapper>
        { children }
      </FormWrapper>
    </Formik>
  )
};
