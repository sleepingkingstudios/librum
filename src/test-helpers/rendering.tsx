import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
