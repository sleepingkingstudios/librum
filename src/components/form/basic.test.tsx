import * as React from 'react';
import {
  Field,
  FormikValues,
} from 'formik';

import { waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { render } from '@test-helpers/rendering';

import { BasicForm } from './basic';
import type {
  OnSubmit,
  SubmitHandler,
} from './types';

describe('<Form />', () => {
  it('should submit the form', async () => {
    const submitHandler: SubmitHandler = jest.fn();
    const onSubmit: OnSubmit =
      (values: FormikValues) => submitHandler(values);

    const { getByRole } = render(
      <BasicForm initialValues={{}} onSubmit={onSubmit}>
        <button type="submit">Submit</button>
      </BasicForm>
    );

    userEvent.click(getByRole('button', { name: 'Submit'}));

    await waitFor(() => {
      expect(submitHandler).toHaveBeenCalledWith({});
    });
  });

  it('should match the snapshot', () => {
    const { asFragment } = render(
      <BasicForm initialValues={{}} onSubmit={jest.fn()}>
        <button type="submit">Submit</button>
      </BasicForm>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('with form inputs', () => {
    const Fields = (): JSX.Element => (
      <React.Fragment>
        <label>
          Launch Site
          <Field id="launchSite" name="launchSite" />
        </label>

        <label>
          Mission Name
          <Field id="mission.name" name="mission.name" />
        </label>
      </React.Fragment>
    );
    const defaultValues = {
      launchSite: '',
      mission: {
        name: '',
      }
    };

    it('should submit the form', async () => {
      const submitHandler: SubmitHandler = jest.fn();
      const onSubmit: OnSubmit =
        (values: FormikValues) => submitHandler(values);

      const { getByRole } = render(
        <BasicForm initialValues={defaultValues} onSubmit={onSubmit}>
          <Fields />

          <button type="submit">Submit</button>
        </BasicForm>
      );

      userEvent.click(getByRole('button', { name: 'Submit'}));

      await waitFor(() => {
        expect(submitHandler).toHaveBeenCalledWith(defaultValues);
      });
    });

    describe('when the user inputs values', () => {
      it('should submit the form', async () => {
        const submitHandler: SubmitHandler = jest.fn();
        const onSubmit: OnSubmit =
          (values: FormikValues) => submitHandler(values);
        const expectedValues = {
          launchSite: 'KSC',
          mission: {
            name: 'Warlock IV',
          },
        };

        const { getByRole, getByLabelText } = render(
          <BasicForm initialValues={defaultValues} onSubmit={onSubmit}>
            <Fields />

            <button type="submit">Submit</button>
          </BasicForm>
        );

        userEvent.type(getByLabelText('Launch Site'), 'KSC');

        userEvent.type(getByLabelText('Mission Name'), 'Warlock IV');

        userEvent.click(getByRole('button', { name: 'Submit'}));

        await waitFor(() => {
          expect(submitHandler).toHaveBeenCalledWith(expectedValues);
        });
      });
    });

    describe('with initial values', () => {
      const initialValues = {
        launchSite: 'KSC',
        mission: {
          name: 'Warlock IV',
        },
      };

      it('should submit the form', async () => {
        const submitHandler: SubmitHandler = jest.fn();
        const onSubmit: OnSubmit =
          (values: FormikValues) => submitHandler(values);

        const { getByRole } = render(
          <BasicForm initialValues={initialValues} onSubmit={onSubmit}>
            <Fields />

            <button type="submit">Submit</button>
          </BasicForm>
        );

        userEvent.click(getByRole('button', { name: 'Submit'}));

        await waitFor(() => {
          expect(submitHandler).toHaveBeenCalledWith(initialValues);
        });
      });
    });
  });
});
