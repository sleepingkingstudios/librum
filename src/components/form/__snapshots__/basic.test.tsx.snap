// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
